### PR TITLE
Instructions and tests to support issue binary for issue #182

### LINKED ISSUE #182: Memory exception for all examples on ARM due to TLS issue
### OS / Environment
Arch Linux ARM (AArch32)

### Python version
Python 2.7.13

### Dependencies
alabaster==0.7.10
appdirs==1.4.3
Babel==2.4.0
capstone==3.0.4
coverage==4.3.4
docutils==0.13.1
imagesize==0.7.1
Jinja2==2.9.6
keystone-engine==0.9.1.post3
-e git+https://github.com/trailofbits/manticore.git@44dff1745d2c5952fb947d13e26b0177bdd6221a#egg=Manticore
MarkupSafe==1.0
nose==1.3.7
packaging==16.8
ply==3.10
pyelftools==0.24
Pygments==2.2.0
pyparsing==2.2.0
pytz==2017.2
requests==2.13.0
six==1.10.0
snowballstemmer==1.2.1
Sphinx==1.5.5
unicorn==1.0.0

### Summary of the problem
A memory exception is thrown for all example executables I've built on ARM

### Actual behavior
```
Command line:
  /usr/bin/manticore ./helloworld
Status:
  Memory Exception: No access writing <0000001c>

================ PROC: 00 ================
Memory:
  0000000000010000-0000000000086000  r x 00000000 ./helloworld
  0000000000095000-0000000000097000  rw  00075000 ./helloworld
  0000000000097000-0000000000098000  rw  00000000 
  00000000bffdf000-00000000c0000000  rwx 00000000 
  00000000f7fff000-00000000f8000000  rw  00000000 
  00000000ffff0000-00000000ffff1000  r x 00000000 
CPU:
INSTRUCTION: 0x00000000000234ac:	str	r6, [r2, r3]
 R0: 0x000000000000000a
 R1: 0x0000000000000000
 R2: 0x0000000000000000
 R3: 0x000000000000001c
 R4: 0x0000000000000009
 R5: 0x00000000bfffff52
 R6: 0x00000000000955f8
 R7: 0x0000000000000055
 R8: 0x0000000000000000
 R9: 0x0000000000000000
R10: 0x0000000000000000
R11: 0x0000000000000000
R12: 0x0000000000096b18
R13: 0x00000000bfffed46
R14: 0x0000000000023860
R15: 0x00000000000234b0
APSR: 0x00000000a0000000
  Instruction: 0x234ac	(str r6, [r2, r3])
```

We can see the issue is that r2 has the value 0. DIssasembling the function in which this occurs, we can determine that the value in R2 was copied from the TLS register:

```
Dump of assembler code for function ptmalloc_init.part.6:
[....]
   0x000234a0 <+36>:    mrc     15, 0, r2, cr13, cr0, {3}
   0x000234a4 <+40>:    ldr     r3, [pc, r3]
   0x000234a8 <+44>:    cmp     r5, r1
   0x000234ac <+48>:    str     r6, [r2, r3]
```

The executable is attached [here](https://github.com/trailofbits/manticore/files/953973/hw.zip).

### PR DESCRIPTION
**Pull request by GarretReece**
Implementation of and tests for the `orn`, `it`, `uadd8`, `sel`, `cbnz`, `rev`, and `sxth` instructions. Update the `add` implementation to support the `add <dest>, <op>` form.

Fixes #182

### CODE REVIEW

File: manticore/core/cpu/arm.py [hunk 1]
```diff
@@ -467,9 +469,21 @@ def shouldCommitFlags(cpu):
         return cpu.instruction.update_flags
 
     def shouldExecuteConditional(cpu):
+        #for the IT instuction, the cc applies to the subsequent instructions,
+        #so the IT instruction should be executed regardless of its cc
+        if cpu.instruction.id == ARM_INS_IT:
+            return True
+
         cc = cpu.instruction.cc
         ret = False
```
Comment (mossberg) [inline manticore/core/cpu/arm.py]: lingering unused `ret`

File: manticore/core/cpu/arm.py [hunk 2]
```diff
@@ -467,9 +469,21 @@ def shouldCommitFlags(cpu):
...
 
+        #support for the it[x[y[z]]] <op> instructions
+        if len(cpu._it_conditional):
```
Comment (mossberg) [inline manticore/core/cpu/arm.py]: do we need the `len` here since `_it_conditional` is a list?

File: tests/test_armv7cpu.py [hunk 3]
```diff
@@ -140,12 +140,40 @@ def wrapper(self):
 
     return instr_dec
 
-def itest_custom_thumb(asm):
-    def instr_dec(custom_func):
-        @wraps(custom_func)
+def itest_thumb(asm):
```
Comment (mossberg) [inline tests/test_armv7cpu.py]: nice!

File: manticore/core/cpu/arm.py [hunk 4]
```diff
@@ -838,6 +906,19 @@ def CLZ(cpu, dest, src):
...
+        opval = op.read()
+        _bytes = list()
+        for i in range(4):
+            _bytes.append(Operators.EXTRACT(opval, i*8, 8))
+        dest.write(Operators.CONCAT(32, *_bytes))
+
+    @instruction
+    def SXTH(cpu, dest, op):
+        _op = op.read()
+        dest.write(Operators.SEXTEND(Operators.EXTRACT(_op, 0, 16), 16, 32))
```
Comment (GarretReece) [inline manticore/core/cpu/arm.py]: The documentation I was referencing didn't mention the rotate option, as it turns out. I added a test, and this works just fine. (I'm completely surprised and delighted.)

Comment (mossberg) [inline manticore/core/cpu/arm.py]: ah, I forgot that `.read()` took care of rotations also. nice :)

File: manticore/core/cpu/arm.py [hunk 5]
```diff
@@ -464,12 +467,26 @@ def _wrap_operands(self, ops):
         return [Armv7Operand(self, op) for op in ops]
 
     def shouldCommitFlags(cpu):
+        #workaround for a capstone bug;
+        #the bug has been fixed the 'master' and 'next' branches of capstone as of 2017-07-31
```
Comment (yan) [inline manticore/core/cpu/arm.py]: Add issue id to comments to ease keeping track of it? And move `return` on next line to its own line?

File: manticore/core/cpu/arm.py [hunk 6]
```diff
@@ -727,8 +783,12 @@ def ADC(cpu, dest, op1, op2):
         return result, carry, overflow
 
     @instruction
-    def ADD(cpu, dest, src, add):
-        result, carry, overflow = cpu._ADD(src.read(), add.read())
+    def ADD(cpu, dest, src, add=None):
+        if add:
+            result, carry, overflow = cpu._ADD(src.read(), add.read())
+        else:
+            '''support for the thumb mode version of adds <dest>, <immediate>'''
```
Comment (yan) [inline manticore/core/cpu/arm.py]: string -> comment?

**Pull request merged**